src/Language/PureScript/Types.hs

```diff
@@ -110,63 +224,65 @@ $(A.deriveJSON A.defaultOptions ''ConstraintData)
 -- |
 -- Convert a row to a list of pairs of labels and types
 --
-rowToList :: Type -> ([(Label, Type)], Type)
-rowToList (RCons name ty row) = let (tys, rest) = rowToList row
-                                in ((name, ty):tys, rest)
+rowToList :: Type a -> ([(Label, Type a)], Type a)
+rowToList (RCons name ty row _) =
+  let (tys, rest) = rowToList row
+  in ((name, ty):tys, rest)
 rowToList r = ([], r)
 
 -- |
 -- Convert a list of labels and types to a row
 --
-rowFromList :: ([(Label, Type)], Type) -> Type
+rowFromList :: ([(Label, Type a)], Type a) -> Type a
 rowFromList ([], r) = r
-rowFromList ((name, t):ts, r) = RCons name t (rowFromList (ts, r))
+rowFromList ((name, t):ts, r) = RCons name t (rowFromList (ts, r)) (extractTypeAnn r)
 
 -- |
 -- Check whether a type is a monotype
 --
-isMonoType :: Type -> Bool
+isMonoType :: Type a -> Bool
 isMonoType ForAll{} = False
-isMonoType (ParensInType t) = isMonoType t
-isMonoType (KindedType t _) = isMonoType t
-isMonoType _        = True
+isMonoType (ParensInType t _) = isMonoType t
+isMonoType (KindedType t _ _) = isMonoType t
+isMonoType _ = True
 
 -- |
 -- Universally quantify a type
 --
-mkForAll :: [Text] -> Type -> Type
-mkForAll args ty = foldl (\t arg -> ForAll arg t Nothing) ty args
+mkForAll :: [Text] -> a -> Type a -> Type a
+mkForAll args a ty = foldl (\t arg -> ForAll arg t Nothing a) ty args
 
 -- |
 -- Replace a type variable, taking into account variable shadowing
 --
-replaceTypeVars :: Text -> Type -> Type -> Type
+replaceTypeVars :: Text -> Type a -> Type a -> Type a
 replaceTypeVars v r = replaceAllTypeVars [(v, r)]
 
 -- |
 -- Replace named type variables with types
 --
-replaceAllTypeVars :: [(Text, Type)] -> Type -> Type
+replaceAllTypeVars :: [(Text, Type a)] -> Type a -> Type a
 replaceAllTypeVars = go []
   where
 
-  go :: [Text] -> [(Text, Type)] -> Type -> Type
-  go _  m (TypeVar v) = fromMaybe (TypeVar v) (v `lookup` m)
-  go bs m (TypeApp t1 t2) = TypeApp (go bs m t1) (go bs m t2)
-  go bs m f@(ForAll v t sco) | v `elem` keys = go bs (filter ((/= v) . fst) m) f
-                             | v `elem` usedVars =
-                               let v' = genName v (keys ++ bs ++ usedVars)
-                                   t' = go bs [(v, TypeVar v')] t
-                               in ForAll v' (go (v' : bs) m t') sco
-                             | otherwise = ForAll v (go (v : bs) m t) sco
+  go :: [Text] -> [(Text, Type a)] -> Type a -> Type a
+  go _  m (TypeVar v a) = fromMaybe (TypeVar v a) (v `lookup` m)
+  go bs m (TypeApp t1 t2 a) = TypeApp (go bs m t1) (go bs m t2) a
+  go bs m f@(ForAll v t sco a)
+    | v `elem` keys = go bs (filter ((/= v) . fst) m) f
+    | v `elem` usedVars =
+        let v' = genName v (keys ++ bs ++ usedVars)
+            t' = go bs [(v, TypeVar v' a)] t
```

Comment:
I got stuck here when I tried this (I didn't get nearly as far). The annotation on `TypeVar v' a` seems wrong, since if the annotations include kinds (for example), then this would have the wrong kind. I think we'd need to take the annotation from the type variable we're replacing.

Probably not a big deal for now though.